File: sdk/python/vanilla_agent.py
#!/usr/bin/env python3
"""
Vanilla Agent — Reference implementation of the Works With Agents OSI Model.
Full 7-layer stack. Single file. ZERO external dependencies beyond Python stdlib.
Copy, run, get a standards-compliant agent. CC BY 4.0.

Usage:
    python vanilla_agent.py --name "reviewer" --purpose "code review"
    python vanilla_agent.py --name "reviewer" --purpose "code review" --demo
"""

import argparse, hashlib, json, os, signal, sys, time, uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Callable, Optional

# ═══════════════════════════════════════════════════════════════════════
# L2 — Identity Protocol (Ed25519 via hashlib — reference only)
# ═══════════════════════════════════════════════════════════════════════

@dataclass
class AgentIdentity:
    """Ed25519 identity for agent signing. Reference: use nacl/pyca for real crypto."""
    agent_id: str
    public_key_hex: str = field(default_factory=lambda: hashlib.sha256(os.urandom(32)).hexdigest())

    def sign(self, payload: dict) -> str:
        """Sign a payload. Reference impl uses SHA-256 of JSON."""
        msg = json.dumps(payload, sort_keys=True).encode()
        return hashlib.sha256(msg + self.public_key_hex.encode()).hexdigest()

    @staticmethod
    def verify(agent_id: str, payload: dict, signature: str, public_key: str) -> bool:
        msg = json.dumps(payload, sort_keys=True).encode()
        return hashlib.sha256(msg + public_key.encode()).hexdigest() == signature

# ═══════════════════════════════════════════════════════════════════════
# L4 — Handoff Protocol
# ═══════════════════════════════════════════════════════════════════════

@dataclass
class HandoffProtocol:
    """Task transfer with complete context. Reference implementation."""

    def create_handoff(self, task_id: str, sender: AgentIdentity, task: dict,
                       quality_checklist: list[str] = None) -> dict:
        payload = {
            "handoff_id": str(uuid.uuid4()),
            "task_id": task_id,
            "sender": {"agent_id": sender.agent_id},
            "context": {
                "task_description": task.get("description", ""),
                "state_snapshot": task.get("state", {}),
                "quality_checklist": quality_checklist or [],
            },
        }
        payload["sender"]["identity_sig"] = sender.sign(payload)
        return payload

    def accept_handoff(self, handoff: dict, receiver: AgentIdentity) -> dict:
        sender_id = handoff.get("sender", {}).get("agent_id", "")
        return {
            "status": "accepted",
            "handoff_id": handoff.get("handoff_id", ""),
            "receiver": receiver.agent_id,
            "estimated_completion": datetime.now(timezone.utc).isoformat(),
        }

# ═══════════════════════════════════════════════════════════════════════
# L5 — Coordination Protocol
# ═══════════════════════════════════════════════════════════════════════

class AgentRole(Enum):
    LEADER = "leader"
    FOLLOWER = "follower"
    CANDIDATE = "candidate"

@dataclass
class CoordinationClient:
    """Leader election, work distribution, conflict resolution."""
    agent_id: str
    role: AgentRole = AgentRole.FOLLOWER
    term: int = 0
    leader_id: Optional[str] = None
    peers: list[str] = field(default_factory=list)
    _work_queue: dict = field(default_factory=dict)
    _conflict_log: dict = field(default_factory=dict)

    def start_election(self) -> dict:
        self.role = AgentRole.CANDIDATE
        self.term += 1
        return {"type": "elect", "candidate": self.agent_id, "term": self.term}

    def become_leader(self) -> dict:
        self.role = AgentRole.LEADER
        self.leader_id = self.agent_id
        return {"type": "heartbeat", "leader": self.agent_id, "term": self.term}

    def assign_work(self, target: str, task: dict, priority: int = 5) -> str:
        tid = f"task-{uuid.uuid4().hex[:8]}"
        self._work_queue[tid] = {"target": target, "task": task, "priority": priority, "status": "assigned"}
        return tid

    def get_status(self) -> dict:
        return {"role": self.role.value, "leader": self.leader_id, "term": self.term,
                "active_tasks": len(self._work_queue)}

# ═══════════════════════════════════════════════════════════════════════
# L7 — Transaction Protocol
# ═══════════════════════════════════════════════════════════════════════

class TransactionStatus(Enum):
    INTENT = "intent"
    COMPLETED = "completed"
    FAILED = "failed"
    ROLLED_BACK = "rolled_back"

@dataclass
class Transaction:
    tx_id: str
    intent: str
    agent_id: str
    status: TransactionStatus = TransactionStatus.INTENT
    created_at: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    completed_at: Optional[str] = None
    result: any = None
    error: Optional[str] = None

class TransactionLedger:
    """Immutable audit trail for agent actions."""
    def __init__(self, agent_id: str):
        self.agent_id = agent_id
        self._ledger: dict[str, Transaction] = {}
        self._completed: set[str] = set()

    def intent(self, description: str, metadata: dict = None) -> Transaction:
        tx = Transaction(tx_id=str(uuid.uuid4()), intent=description, agent_id=self.agent_id)
        self._ledger[tx.tx_id] = tx
        return tx

    def execute(self, tx: Transaction, fn: Callable) -> Transaction:
        try:
            tx.result = fn()
            tx.status = TransactionStatus.COMPLETED
            tx.completed_at = datetime.now(timezone.utc).isoformat()
        except Exception as e:
            tx.status = TransactionStatus.FAILED
            tx.error = str(e)
        return tx

    def audit(self) -> list[dict]:
        return [{"id": t.tx_id, "intent": t.intent, "status": t.status.value,
                 "created": t.created_at, "error": t.error} for t in self._ledger.values()]

    def stats(self) -> dict:
        txs = list(self._ledger.values())
        return {"total": len(txs), "completed": sum(1 for t in txs if t.status == TransactionStatus.COMPLETED),
                "failed": sum(1 for t in txs if t.status == TransactionStatus.FAILED)}

# ═══════════════════════════════════════════════════════════════════════
# L3 — Capability Manifest
# ═══════════════════════════════════════════════════════════════════════

@dataclass
class CapabilityManifest:
    agent_id: str
    name: str
    purpose: str
    tools: list[str]
    model: str = "vanilla"
    max_tokens: int = 131072
    trust_score: float = 50.0
    deployment: str = "local"

    def to_dict(self) -> dict:
        return {k: v for k, v in self.__dict__.items()}

# ═══════════════════════════════════════════════════════════════════════
# L7 — Compliance-as-Code (simplified for standalone)
# ═══════════════════════════════════════════════════════════════════════

class ComplianceGate:
    """Simple compliance rule engine. Extend with real regulation packs."""

    DEFAULT_RULES = {
        "nhs_dtac": [
            {"field": "data_classification", "operator": "not_in",
             "value": ["patient_identifiable", "clinical_confidential"],
             "message": "Patient data must not be processed without DPIA"},
        ],
        "gdpr": [
            {"field": "data_classification", "operator": "not_in",
             "value": ["personal_data_unconsented"],
             "message": "Unconsented personal data blocked"},
        ],
    }

    @classmethod
    def validate(cls, regulation: str, action: dict) -> dict:
        rules = cls.DEFAULT_RULES.get(regulation, [])
        violations = []
        for rule in rules:
            actual = action.get(rule["field"], "")
            if rule["operator"] == "not_in" and actual in rule["value"]:
                violations.append(rule["message"])
        return {"passed": len(violations) == 0, "violations": violations}

# ═══════════════════════════════════════════════════════════════════════
# Vanilla Agent — 7-Layer OSI Reference
# ═══════════════════════════════════════════════════════════════════════

class VanillaAgent:
    """A standards-compliant AI agent implementing the full 7-layer OSI Model."""

    def __init__(self, name: str, purpose: str, tools: list[str] = None):
        self.name = name
        self.purpose = purpose
        self.agent_id = f"agent-{name}-{uuid.uuid4().hex[:6]}"
        self.tools = tools or ["terminal", "file", "web"]
        self.started_at = datetime.now(timezone.utc).isoformat()
        self._running = False

        # L2: Identity
        self.identity = AgentIdentity(self.agent_id)
        # L3: Capabilities
        self.capabilities = CapabilityManifest(self.agent_id, name, purpose, self.tools)
        # L4: Handoff
        self.handoff = HandoffProtocol()
        self._handoff_queue: list[dict] = []
        # L5: Coordination
        self.coord = CoordinationClient(self.agent_id)
        # L6: Task tracking
        self.tasks_done = 0
        self.tasks_failed = 0
        # L7: Governance
        self.ledger = TransactionLedger(self.agent_id)

    # ── Bootstrap ────────────────────────────────────────────────────
    def boot(self) -> dict:
        """L1-L3: Generate identity and publish capabilities."""
        return {
            "agent_id": self.agent_id,
            "identity": self.identity.public_key_hex[:16] + "...",
            "capabilities": self.capabilities.to_dict(),
            "status": "ready",
        }

    # ── L4: Handoff ──────────────────────────────────────────────────
    def receive_task(self, task: dict) -> dict:
        """Accept a task from another agent via Handoff Protocol."""
        task_id = task.get("task_id", str(uuid.uuid4()))
        handoff = self.handoff.create_handoff(
            task_id, self.identity, task,
            quality_checklist=task.get("checklist", ["Verify output", "Run tests"]),
        )
        response = self.handoff.accept_handoff(handoff, self.identity)
        self._handoff_queue.append({"handoff": handoff, "accepted": response["status"]})

        # Execute with compliance + audit
        result = self.execute(task)
        return {**response, "result": result}

    def handoff_to(self, target_agent: str, task: dict) -> dict:
        """Transfer a task to another agent."""
        task_id = str(uuid.uuid4())
        handoff = self.handoff.create_handoff(task_id, self.identity, task)
        return {"handoff_id": handoff["handoff_id"], "target": target_agent, "task_id": task_id}

    # ── L5: Coordination ─────────────────────────────────────────────
    def join_fleet(self, peers: list[str] = None) -> dict:
        """Join a fleet and coordinate."""
        self.coord.peers = peers or []
        self.coord.start_election()
        return self.coord.get_status()

    # ── L6-L7: Execute with compliance gates ─────────────────────────
    def execute(self, task: dict) -> dict:
        """Execute a task with compliance validation and audit trail."""

        # L7: Compliance check before execution
        for regulation in ["nhs_dtac", "gdpr"]:
            result = ComplianceGate.validate(regulation, task)
            if not result["passed"]:
                return {"status": "blocked", "reason": "compliance",
                        "violations": result["violations"]}

        # L7: Transaction intent → execution → confirmation
        tx = self.ledger.intent(task.get("description", "execute task"))

        try:
            ok = self._run(task)
            self.ledger.execute(tx, lambda: {"ok": ok})
            self.tasks_done += 1
            status = "completed"
        except Exception as e:
            self.ledger.execute(tx, lambda: (_ for _ in ()).throw(e))
            self.tasks_failed += 1
            status = "failed"

        return {"status": status, "tx_id": tx.tx_id,
                "tasks_done": self.tasks_done, "tasks_failed": self.tasks_failed}

    def _run(self, task: dict) -> bool:
        """Simulate tool execution. Replace with actual tool calls."""
        tool = task.get("tool", "echo")
        print(f"  [{self.agent_id}] Running {tool}...")
        return True

    # ── Stats ────────────────────────────────────────────────────────
    def report(self) -> dict:
        return {
            "agent_id": self.agent_id,
            "name": self.name,
            "purpose": self.purpose,
            "tools": self.tools,
            "uptime_s": int((datetime.now(timezone.utc) -
                             datetime.fromisoformat(self.started_at)).total_seconds()),
            "tasks_done": self.tasks_done,
            "tasks_failed": self.tasks_failed,
            "handoffs_received": len(self._handoff_queue),
            "coordination": self.coord.get_status(),
            "audit": self.ledger.stats(),
            "identity": self.identity.public_key_hex[:16] + "...",
        }

    # ── Run loop ─────────────────────────────────────────────────────
    def start(self):
        self._running = True
        boot = self.boot()
        print(f"╔══════════════════════════════════╗")
        print(f"║  {self.agent_id}  ║")
        print(f"║  {self.name:28s}  ║")
        print(f"║  {self.purpose:28s}  ║")
        print(f"║  ID: {boot['identity']:21s}  ║")
        print(f"║  Tools: {', '.join(self.tools):18s}  ║")
        print(f"╚══════════════════════════════════╝")
        print()

        while self._running:
            time.sleep(1)

    def stop(self):
        self._running = False


# ═══════════════════════════════════════════════════════════════════════
# Demo
# ═══════════════════════════════════════════════════════════════════════

def demo():
    """Run a full 7-layer demo with two agents coordinating."""
    print("=" * 60)
    print("  Vanilla Agent — OSI Reference Implementation")
    print("  Works With Agents · CC BY 4.0")
    print("=" * 60)
    print()

    # Create two agents
    builder = VanillaAgent("builder", "build software", ["terminal", "file", "git"])
    reviewer = VanillaAgent("reviewer", "review code", ["terminal", "file", "web"])

    # Boot
    print("── L1-L3: Boot, Identity, Capabilities ──")
    b_boot = builder.boot()
    r_boot = reviewer.boot()
    print(f"  {b_boot['agent_id']} ready — {b_boot['capabilities']['tools']}")
    print(f"  {r_boot['agent_id']} ready — {r_boot['capabilities']['tools']}")
    print()

    # Handoff
    print("── L4: Handoff Protocol ──")
    task = {"task_id": "task-001", "description": "Build API endpoint",
            "tool": "file", "data_classification": "internal",
            "checklist": ["Tests pass", "No secrets in code"]}
    result = reviewer.receive_task(task)
    print(f"  {reviewer.agent_id} received task: {result['status']}")
    print(f"  TxID: {result.get('result', {}).get('tx_id', '')}")
    print()

    # Coordination
    print("── L5: Coordination Protocol ──")
    fleet = builder.join_fleet([reviewer.agent_id])
    print(f"  {builder.agent_id} fleet: {fleet['role']} (term {fleet['term']})")

    # Assign work via leader
    task_id = builder.coord.assign_work(reviewer.agent_id,
        {"description": "Review PR #42", "tool": "terminal", "data_classification": "internal"})
    print(f"  {builder.agent_id} assigned {task_id} to {reviewer.agent_id}")
    print()

    # Execute with compliance
    print("── L6-L7: Execute + Governance ──")
    # Safe task
    result = builder.execute({"description": "Run unit tests", "tool": "terminal",
                              "data_classification": "internal"})
    print(f"  Safe task: {result['status']}")

    # Blocked task
    result = builder.execute({"description": "Process patient data", "tool": "file",
                              "data_classification": "patient_identifiable"})
    print(f"  Blocked task: {result['status']} — {result.get('violations', [])}")
    print()

    # Audit trail
    print("── L7: Audit Trail ──")
    for entry in builder.ledger.audit():
        print(f"  [{entry['status']}] {entry['intent']}")
    print()

    # Reports
    print("── Agent Reports ──")
    for agent in [builder, reviewer]:
        r = agent.report()
        print(f"  {r['agent_id']}: {r['tasks_done']} done, {r['tasks_failed']} failed, "
              f"{len(agent._handoff_queue)} handoffs")

    print()
    print("=" * 60)
    print("  Demo complete. 7 layers, 2 agents, 0 external deps.")
    print("  pip install workswithagents  →  full SDK with 12 protocols")
    print("=" * 60)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Vanilla Agent — OSI Reference Implementation")
    parser.add_argument("--name", default="agent", help="Agent name")
    parser.add_argument("--purpose", default="general-purpose", help="Agent purpose")
    parser.add_argument("--tools", default="terminal,file", help="Comma-separated tools")
    parser.add_argument("--demo", action="store_true", help="Run full 7-layer demo")
    args = parser.parse_args()

    if args.demo:
        demo()
    else:
        agent = VanillaAgent(
            name=args.name,
            purpose=args.purpose,
            tools=[t.strip() for t in args.tools.split(",")],
        )
        def handle(sig, frame):
            agent.stop()
            sys.exit(0)
        signal.signal(signal.SIGINT, handle)
        signal.signal(signal.SIGTERM, handle)
        agent.start()

File: sdk/typescript/vanilla_agent.ts
#!/usr/bin/env node
/**
 * Vanilla Agent — TypeScript reference implementation.
 * Full 7-layer OSI Model. Single file. Zero dependencies.
 * Copy, run, get a standards-compliant agent. CC BY 4.0.
 *
 * Usage:
 *   npx tsx vanilla_agent.ts --name "reviewer" --purpose "code review"
 *   npx tsx vanilla_agent.ts --demo
 */

import * as crypto from "crypto";

// ═══════════════════════════════════════════════════════════════════════
// L2 — Identity Protocol
// ═══════════════════════════════════════════════════════════════════════

class AgentIdentity {
  publicKeyHex: string;

  constructor(public agentId: string) {
    this.publicKeyHex = crypto.createHash("sha256")
      .update(crypto.randomBytes(32)).digest("hex");
  }

  sign(payload: Record<string, unknown>): string {
    const msg = JSON.stringify(payload, Object.keys(payload).sort());
    return crypto.createHash("sha256")
      .update(msg + this.publicKeyHex).digest("hex");
  }

  static verify(agentId: string, payload: Record<string, unknown>,
                signature: string, publicKey: string): boolean {
    const msg = JSON.stringify(payload, Object.keys(payload).sort());
    return crypto.createHash("sha256")
      .update(msg + publicKey).digest("hex") === signature;
  }
}

// ═══════════════════════════════════════════════════════════════════════
// L4 — Handoff Protocol
// ═══════════════════════════════════════════════════════════════════════

interface HandoffPayload {
  handoff_id: string;
  task_id: string;
  sender: { agent_id: string; identity_sig?: string };
  context: { task_description: string; state_snapshot: Record<string, unknown>; quality_checklist: string[] };
}

class HandoffProtocol {
  createHandoff(taskId: string, sender: AgentIdentity,
                task: Record<string, unknown>, checklist: string[] = []): HandoffPayload {
    const payload: HandoffPayload = {
      handoff_id: crypto.randomUUID(),
      task_id: taskId,
      sender: { agent_id: sender.agentId },
      context: {
        task_description: (task.description as string) || "",
        state_snapshot: (task.state as Record<string, unknown>) || {},
        quality_checklist: checklist.length ? checklist : ["Verify output", "Run tests"],
      },
    };
    payload.sender.identity_sig = sender.sign(payload as unknown as Record<string, unknown>);
    return payload;
  }

  acceptHandoff(handoff: HandoffPayload, receiver: AgentIdentity) {
    return {
      status: "accepted",
      handoff_id: handoff.handoff_id,
      receiver: receiver.agentId,
      estimated_completion: new Date().toISOString(),
    };
  }
}

// ═══════════════════════════════════════════════════════════════════════
// L5 — Coordination Protocol
// ═══════════════════════════════════════════════════════════════════════

enum AgentRole { LEADER = "leader", FOLLOWER = "follower", CANDIDATE = "candidate" }

class CoordinationClient {
  role: AgentRole = AgentRole.FOLLOWER;
  term = 0;
  leaderId: string | null = null;
  workQueue: Map<string, Record<string, unknown>> = new Map();

  constructor(public agentId: string, public peers: string[] = []) {}

  startElection() {
    this.role = AgentRole.CANDIDATE;
    this.term += 1;
    return { type: "elect", candidate: this.agentId, term: this.term };
  }

  becomeLeader() {
    this.role = AgentRole.LEADER;
    this.leaderId = this.agentId;
    return { type: "heartbeat", leader: this.agentId, term: this.term };
  }

  assignWork(target: string, task: Record<string, unknown>, priority = 5): string {
    const tid = `task-${crypto.randomUUID().slice(0, 8)}`;
    this.workQueue.set(tid, { target, task, priority, status: "assigned" });
    return tid;
  }

  getStatus() {
    return { role: this.role, leader: this.leaderId, term: this.term, active_tasks: this.workQueue.size };
  }
}

// ═══════════════════════════════════════════════════════════════════════
// L7 — Transaction Protocol
// ═══════════════════════════════════════════════════════════════════════

enum TxStatus { INTENT = "intent", COMPLETED = "completed", FAILED = "failed", ROLLED_BACK = "rolled_back" }

interface Transaction {
  tx_id: string; intent: string; agent_id: string;
  status: TxStatus; created_at: string; completed_at: string | null;
  result: unknown; error: string | null;
}

class TransactionLedger {
  private ledger: Map<string, Transaction> = new Map();

  constructor(public agentId: string) {}

  intent(description: string): Transaction {
    const tx: Transaction = {
      tx_id: crypto.randomUUID(), intent: description, agent_id: this.agentId,
      status: TxStatus.INTENT, created_at: new Date().toISOString(),
      completed_at: null, result: null, error: null,
    };
    this.ledger.set(tx.tx_id, tx);
    return tx;
  }

  execute(tx: Transaction, fn: () => unknown): Transaction {
    try {
      tx.result = fn();
      tx.status = TxStatus.COMPLETED;
      tx.completed_at = new Date().toISOString();
    } catch (e) {
      tx.status = TxStatus.FAILED;
      tx.error = e instanceof Error ? e.message : String(e);
    }
    return tx;
  }

  audit() {
    return Array.from(this.ledger.values()).map(t => ({
      id: t.tx_id, intent: t.intent, status: t.status,
      created: t.created_at, error: t.error,
    }));
  }

  stats() {
    const txs = Array.from(this.ledger.values());
    return { total: txs.length, completed: txs.filter(t => t.status === TxStatus.COMPLETED).length,
             failed: txs.filter(t => t.status === TxStatus.FAILED).length };
  }
}

// ═══════════════════════════════════════════════════════════════════════
// L7 — Compliance Gate
// ═══════════════════════════════════════════════════════════════════════

const COMPLIANCE_RULES: Record<string, Array<{ field: string; operator: string; value: string[]; message: string }>> = {
  nhs_dtac: [
    { field: "data_classification", operator: "not_in",
      value: ["patient_identifiable", "clinical_confidential"],
      message: "Patient data must not be processed without DPIA" },
  ],
  gdpr: [
    { field: "data_classification", operator: "not_in",
      value: ["personal_data_unconsented"],
      message: "Unconsented personal data blocked" },
  ],
};

function validateCompliance(regulation: string, action: Record<string, unknown>) {
  const rules = COMPLIANCE_RULES[regulation] || [];
  const violations: string[] = [];
  for (const rule of rules) {
    const actual = String(action[rule.field] || "");
    if (rule.operator === "not_in" && rule.value.includes(actual)) {
      violations.push(rule.message);
    }
  }
  return { passed: violations.length === 0, violations };
}

// ═══════════════════════════════════════════════════════════════════════
// Vanilla Agent
// ═══════════════════════════════════════════════════════════════════════

class VanillaAgent {
  agentId: string;
  identity: AgentIdentity;
  capabilities: Record<string, unknown>;
  handoff: HandoffProtocol;
  coord: CoordinationClient;
  ledger: TransactionLedger;
  handoffQueue: Array<{ handoff: HandoffPayload; accepted: string }> = [];
  tasksDone = 0;
  tasksFailed = 0;
  startedAt: string;

  constructor(public name: string, public purpose: string, public tools: string[] = ["terminal", "file"]) {
    this.agentId = `agent-${name}-${crypto.randomUUID().slice(0, 6)}`;
    this.startedAt = new Date().toISOString();
    this.identity = new AgentIdentity(this.agentId);
    this.handoff = new HandoffProtocol();
    this.coord = new CoordinationClient(this.agentId);
    this.ledger = new TransactionLedger(this.agentId);
    this.capabilities = { agent_id: this.agentId, name, purpose, tools, model: "vanilla", trust_score: 50 };
  }

  boot() {
    return { agent_id: this.agentId, capabilities: this.capabilities, status: "ready" };
  }

  receiveTask(task: Record<string, unknown>) {
    const taskId = (task.task_id as string) || crypto.randomUUID();
    const handoff = this.handoff.createHandoff(taskId, this.identity, task,
      (task.checklist as string[]) || []);
    const response = this.handoff.acceptHandoff(handoff, this.identity);
    this.handoffQueue.push({ handoff, accepted: response.status });
    const result = this.execute(task);
    return { ...response, result };
  }

  execute(task: Record<string, unknown>) {
    for (const regulation of ["nhs_dtac", "gdpr"]) {
      const result = validateCompliance(regulation, task);
      if (!result.passed) return { status: "blocked", violations: result.violations };
    }

    const tx = this.ledger.intent((task.description as string) || "execute task");
    try {
      const tool = (task.tool as string) || "echo";
      console.log(`  [${this.agentId}] Running ${tool}...`);
      this.ledger.execute(tx, () => ({ ok: true }));
      this.tasksDone++;
      return { status: "completed", tx_id: tx.tx_id };
    } catch (e) {
      this.ledger.execute(tx, () => { throw e; });
      this.tasksFailed++;
      return { status: "failed", error: String(e) };
    }
  }

  joinFleet(peers: string[] = []) {
    this.coord.peers = peers;
    this.coord.startElection();
    return this.coord.getStatus();
  }

  report() {
    return {
      agent_id: this.agentId, name: this.name, purpose: this.purpose,
      tools: this.tools, tasks_done: this.tasksDone, tasks_failed: this.tasksFailed,
      handoffs_received: this.handoffQueue.length,
      coordination: this.coord.getStatus(), audit: this.ledger.stats(),
    };
  }
}

// ═══════════════════════════════════════════════════════════════════════
// Demo
// ═══════════════════════════════════════════════════════════════════════

async function demo() {
  console.log("═".repeat(60));
  console.log("  Vanilla Agent — OSI Reference (TypeScript)");
  console.log("  Works With Agents · CC BY 4.0");
  console.log("═".repeat(60) + "\n");

  const builder = new VanillaAgent("builder", "build software", ["terminal", "file", "git"]);
  const reviewer = new VanillaAgent("reviewer", "review code", ["terminal", "file", "web"]);

  console.log("── L1-L3: Boot, Identity, Capabilities ──");
  console.log(`  ${builder.boot().agent_id} ready`);
  console.log(`  ${reviewer.boot().agent_id} ready\n`);

  console.log("── L4: Handoff Protocol ──");
  const result = reviewer.receiveTask({
    task_id: "task-001", description: "Build API endpoint",
    tool: "file", data_classification: "internal",
  });
  console.log(`  ${reviewer.agentId} received: ${result.status}\n`);

  console.log("── L5: Coordination Protocol ──");
  const fleet = builder.joinFleet([reviewer.agentId]);
  console.log(`  ${builder.agentId} fleet: ${fleet.role} (term ${fleet.term})`);
  builder.coord.assignWork(reviewer.agentId, { description: "Review PR", tool: "terminal", data_classification: "internal" });
  console.log(`  ${builder.agentId} assigned work to ${reviewer.agentId}\n`);

  console.log("── L6-L7: Execute + Governance ──");
  const safe = builder.execute({ description: "Run tests", tool: "terminal", data_classification: "internal" });
  console.log(`  Safe task: ${safe.status}`);
  const blocked = builder.execute({ description: "Process patient data", tool: "file", data_classification: "patient_identifiable" });
  console.log(`  Blocked: ${blocked.status} — ${(blocked.violations as string[] || []).join(", ")}\n`);

  console.log("── L7: Audit Trail ──");
  for (const entry of builder.ledger.audit()) {
    console.log(`  [${entry.status}] ${entry.intent}`);
  }
  console.log();

  console.log("── Reports ──");
  for (const agent of [builder, reviewer]) {
    const r = agent.report();
    console.log(`  ${r.agent_id}: ${r.tasks_done} done, ${r.tasks_failed} failed`);
  }
  console.log("\n" + "═".repeat(60));
  console.log("  Demo complete. 7 layers, 2 agents, 0 external deps.");
  console.log("═".repeat(60));
}

// CLI
const args = process.argv.slice(2);
if (args.includes("--demo")) {
  demo().catch(console.error);
} else {
  console.log("Vanilla Agent — TypeScript. Use --demo for the full OSI demo.");
}
